swarm/bmt/bmt.go

```diff
@@ -244,130 +257,148 @@ func (t *Tree) Draw(hash []byte, d int) string {
+			nodes[i] = newNode(i, parent)
 		}
 		prevlevel = nodes
-		level++
 		count *= 2
 	}
-	// the datanode level is the nodes on the last level where
-	return &Tree{
-		leaves: prevlevel,
+	// the datanode level is the nodes on the last level
+	return &tree{
+		leaves:  prevlevel,
+		result:  make(chan []byte, 1),
+		segment: make([]byte, segmentSize),
+		section: make([]byte, 2*segmentSize),
 	}
 }
 
 // methods needed by hash.Hash
 
 // Size returns the size
 func (h *Hasher) Size() int {
-	return h.size
+	return h.pool.SegmentSize
 }
 
 // BlockSize returns the block size
 func (h *Hasher) BlockSize() int {
-	return h.blocksize
+	return h.pool.SegmentSize
+}
+
+// Hash hashes the data and the span using the bmt hasher
+func Hash(h *Hasher, span, data []byte) []byte {
+	h.ResetWithLength(span)
+	h.Write(data)
+	return h.Sum(nil)
+}
+
+// Datalength returns the maximum data size that is hashed by the hasher =
+// segment count times segment size
+func (h *Hasher) DataLength() int {
+	return h.pool.Datalength
 }
 
 // Sum returns the hash of the buffer
 // hash.Hash interface Sum method appends the byte slice to the underlying
 // data before it calculates and returns the hash of the chunk
+// caller must make sure Sum is not called concurrently with Write or WriteSegment
 func (h *Hasher) Sum(b []byte) (r []byte) {
+	return h.sum(b, true, true)
+}
+
+// sum implements Sum taking parameters
+// * if the tree is released right away
+// * if sequential write is used (can read sections)
+func (h *Hasher) sum(b []byte, release, section bool) (r []byte) {
 	t := h.bmt
-	i := h.cur
-	n := t.leaves[i]
-	j := i
-	// must run strictly before all nodes calculate
-	// datanodes are guaranteed to have a parent
-	if len(h.segment) > h.size && i > 0 && n.parent != nil {
-		n = n.parent
+	h.finalise(section)
+	if t.offset > 0 { // get the last node (double segment)
+
+		// padding the segment  with zero
+		copy(t.segment[t.offset:], h.pool.zerohashes[0])
+	}
+	if section {
+		if t.cur%2 == 1 {
+			// if just finished current segment, copy it to the right half of the chunk
+			copy(t.section[h.pool.SegmentSize:], t.segment)
+		} else {
+			// copy segment to front of section, zero pad the right half
+			copy(t.section, t.segment)
+			copy(t.section[h.pool.SegmentSize:], h.pool.zerohashes[0])
+		}
+		h.writeSection(t.cur, t.section)
 	} else {
-		i *= 2
+		h.writeSegment(t.cur, t.segment)
+	}
+	bmtHash := <-t.result
+	span := t.span
+
+	if release {
+		h.releaseTree()
 	}
-	d := h.finalise(n, i)
-	h.writeSegment(j, h.segment, d)
-	c := <-h.result
-	h.releaseTree()
-
-	// sha3(length + BMT(pure_chunk))
-	if h.blockLength == nil {
-		return c
+	// sha3(span + BMT(pure_chunk))
+	if span == nil {
+		return bmtHash
 	}
-	res := h.pool.hasher()
-	res.Reset()
-	res.Write(h.blockLength)
-	res.Write(c)
-	return res.Sum(nil)
+	bh := h.pool.hasher()
+	bh.Reset()
+	bh.Write(span)
+	bh.Write(bmtHash)
+	return bh.Sum(b)
 }
 
 // Hasher implements the SwarmHash interface
```

Comment:
for this needs to put `Hash` function back but we need `WriteSegment` , `NewProof` ,`VerifyProof` functions